bluetooth: tester: LE Audio: Fix sending ISO data

mkasenberg opened the pull request:
The host app should increment the sequence number within SDU interval whether it has any data to send data to the controller or not.

mkasenberg committed:
bluetooth: tester: LE Audio: Fix sending ISO data

The host app should increment the sequence number within SDU interval
whether it has any data to send data to the controller or not.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -525,6 +539,22 @@ static void stream_started(struct bt_bap_stream *stream)
+			   K_THREAD_STACK_SIZEOF(iso_data_thread_stack_area),
+			   ISO_DATA_THREAD_PRIORITY, NULL);
+	k_work_init_delayable(&a_stream->audio_send_work, audio_send_timeout);
+	k_work_schedule_for_queue(&iso_data_work_q, &a_stream->audio_send_work,
+				  K_USEC(a_stream->stream.qos->interval));
```
One important part here is that the seq_num starts at 0 when the CIS is connected, and since you schedule the `audio_send_work` here with a delay, it should probably be incremented before first TX. 

Alternatively, use `K_NO_WAIT` when starting the first TX here

Thalley replied:
@mkasenberg is there a reason to not start sending immediately?

mkasenberg replied:
The seq_num is incremented in `audio_clock_work` and TX is perforemed by `audio_send_work`. But just in case, let's start seq_num from 1:
```c
a_stream->seq_num = 1;
```

mkasenberg replied:
Sorry, did not refreshed the site. I was not sure

> @mkasenberg is there a reason to not start sending immediately?

I was not sure if I can rely on priority of system work queue.

Thalley replied:
> I was not sure if I can rely on priority of system work queue.

That is a fair point :)

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,93 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+	if (size != 0) {
+		net_buf_add_mem(buf, data, size);
+
+		/* Free ring buffer memory */
+		err = ring_buf_get_finish(&audio_ring_buf, size);
+		if (err != 0) {
+			LOG_ERR("Error freeing ring buffer memory: %d", err);
+		}
```
Can we really free this before calling `bt_bap_stream_send`?

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,93 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+
+	err = bt_bap_stream_send(&stream->stream, buf, seq_num, BT_ISO_TIMESTAMP_NONE);
+
+	if (err != 0) {
+		LOG_ERR("Failed to send audio data to stream: ase_id %d dir seq %d %d err %d",
+			stream->ase_id, stream->stream.dir, seq_num, err);
+	}
```
In case of error we should call `net_buf_unref`

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,93 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+		size, seq_num);
+
+	err = bt_bap_stream_send(&stream->stream, buf, seq_num, BT_ISO_TIMESTAMP_NONE);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,93 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+	/* Get buffer within a ring buffer memory */
+	size = ring_buf_get_claim(&audio_ring_buf, &data, stream->stream.qos->sdu);
+
+	buf = net_buf_alloc(&tx_pool, K_FOREVER);
+	if (!buf) {
```
Oh, I see that this is actually using a different queue than the system workqueue, but it still shouldn't use K_FOREVER IMO :)

mkasenberg commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,93 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+		iso_chan = bt_bap_stream_iso_chan_get(&stream->stream);
+		err = bt_iso_chan_get_tx_sync(iso_chan, &info);
+
+		if (err != 0) {
+			LOG_DBG("Failed to get last seq num: err %d", err);
+		} else {
+			LOG_DBG("bt_bap_stream_iso_chan_get: seq num: %u", info.seq_num);
+		}
+	}
```
I tried to implement some synchronization trick with `bt_iso_chan_get_tx_sync`, but it always returns a sequence number used in with `bt_bap_stream_send`, even though in logs of the Bluetooth Protocol Viewer after about 150-200 ISO packets (empty or not) the PTS controller stops to report received ISO data. I am not sure yet if it is just a BPV behavior, the test case implementation or the Zephyr controller drops the packets.

mkasenberg replied:
[BAP_UCL_STR_BV_033_C_2023_06_09_13_49_33.cfa.txt](https://github.com/zephyrproject-rtos/zephyr/files/11706113/BAP_UCL_STR_BV_033_C_2023_06_09_13_49_33.cfa.txt)

Thalley replied:
There are some confusion about `bt_iso_chan_get_tx_sync` (and some errata open for it) on whether it returns the seq_num of the next SDU to be transmitted or the last SDU that was transmitted. In any case, the `seq_num` we provide to `bt_bap_stream_send` should always be at least 1 larger than the `info.seq_num`

@cvinayak may have additional input here

mkasenberg replied:
If `bt_iso_chan_get_tx_sync` would work, I was thinking about something like this to put here:
```c
	if (wait_counter > 0) {
		wait_counter--;
		return;
	}

	if (seq_num % 201 == 200 || clock_ahead == true) {
		struct bt_iso_chan *iso_chan = bt_bap_stream_iso_chan_get(&audio_stream->stream);
		struct bt_iso_tx_info info;
		err = bt_iso_chan_get_tx_sync(iso_chan, &info);

		if (err != 0) {
			LOG_DBG("Failed to get last seq num: err %d", err);
			return;
		} else {
			LOG_DBG("bt_bap_stream_iso_chan_get: seq num: %u", info.seq_num);
		}

		uint16_t seq_num_diff = audio_stream->seq_num - info.seq_num;
		if (seq_num_diff > 1) {
			/* Host clock desynchronized with Controller clock */
			if (clock_ahead == false) {
				/* Assume Host clock is ahead and wait few intervals */
				clock_ahead = true;
				wait_counter = seq_num_diff - 1;
				return;
			} else {
				/* Assume the Host clock is late and shift forward the sequence number */
				audio_stream->seq_num += 2 * seq_num_diff + 1;
			}
		}
	}

	clock_ahead = false;
```

Thalley replied:
Yeah, that makes sense. 

Another thing to check is whether `info.seq_num` is the same as last; if it is that would indicate that the last TX attempt by the host did not go through. The TX sync seq_num is the only way for the host to verify that a TX request was accepted by the controller

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -83,6 +86,17 @@ static struct net_buf_simple *rx_ev_buf = NET_BUF_SIMPLE(CONFIG_BT_ISO_RX_MTU +
+#define ISO_DATA_THREAD_PRIORITY -7
+K_THREAD_STACK_DEFINE(iso_data_thread_stack_area, ISO_DATA_THREAD_STACK_SIZE);
+static struct k_work_q iso_data_work_q;
+
+
+
```
```suggestion

```

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,95 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+	uint32_t size;
+	uint8_t *data;
+	int err;
+	static uint16_t seq_num = 1;
```
Can you elaborate on why we have a `static` `seq_num` in this function, rather than just always getting it from `atomic_get(&stream->seq_num);`?

mkasenberg replied:
I would like to compare here
https://github.com/zephyrproject-rtos/zephyr/blob/1e84b740f9a5c6e04706ac9abf4ce3df5744eac9/tests/bluetooth/tester/src/btp_bap.c#L888
the last accepted sequence number, loaded into `info.seq_num`, with the last requested sequence number. Since the `stream->seq_num` is incremented in a different thread and the `bt_iso_chan_get_tx_sync` can take some time, I need to use `atomic_get(&stream->seq_num)` just before calling the `bt_bap_stream_send`.
At first, I thought that a static variable in a function could hold the last requested sequence number for me. However, I forgot that there will be more streams, so the placement is wrong. Therefore, I will move it into a `last_req_seq_num` field of the `struct audio_stream`.

Thalley replied:
Ah, yes, that makes sense. Perhaps we could store the `last_sen_num` in the `stream` object as well, if we ever want to expand this to support more than 1 stream?

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +853,95 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+	if (err != 0) {
+		LOG_ERR("Failed to send audio data to stream: ase_id %d dir seq %d %d err %d",
+			stream->ase_id, stream->stream.dir, seq_num, err);
+		net_buf_unref(buf);
+	} else if (size != 0) {
+		/* Free ring buffer memory */
+		err = ring_buf_get_finish(&audio_ring_buf, size);
+		if (err != 0) {
+			LOG_ERR("Error freeing ring buffer memory: %d", err);
+		}
+	}
```
if `bt_bap_stream_send` fails, we should still call `ring_buf_get_finish`, right?

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -819,16 +854,99 @@ static uint8_t bap_discover(const void *cmd, uint16_t cmd_len,
+	LOG_DBG("Sending data to ASE: ase_id %d len %d seq %d", stream->stream.ep->status.id,
+		size, stream->last_req_seq_num);
+
+	err = bt_bap_stream_send(&stream->stream, buf, stream->last_req_seq_num,
+				 BT_ISO_TIMESTAMP_NONE);
```
what is the value range of `atomic_t`? The `seq_num` in `bt_bap_stream_send` shall be 16-bit, so suggest to either cast to `uint16_t` to make it explicit that we allow overflows, or reset `stream->last_req_seq_num` to 0 when it reaches `UINT16_MAX`

Thalley reviewed (dismissed):
Changes look good, but suggest to make seq_num explicit 16-bit

Thalley commented on tests/bluetooth/tester/src/btp_bap.c:
```diff
@@ -525,6 +541,20 @@ static void stream_started(struct bt_bap_stream *stream)
 	LOG_DBG("Started stream %p", stream);
+
+	if (stream->dir == BT_AUDIO_DIR_SINK) {
+		a_stream->seq_num = 1;
```
We should add a comment on why this starts at 1 instead of 0 (because the first TX is delayed by 1 interval)

mkasenberg committed:
bluetooth: tester: has: Update name of bt_has_register_param

The bt_has_register_param was renamed to bt_has_features_param
recently. Fixes build errors.

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

mkasenberg committed:
bluetooth: tester: vcp: Remove unused include

Fixes build error:
mesh/keys.h:40:2: error: #error "Crypto library has not been chosen"

Signed-off-by: Magdalena Kasenberg <magdalena.kasenberg@codecoup.pl>

mkasenberg commented:
I rebased and fixed some build errors introduced after recent updates.

cvinayak commented:
@mkasenberg does this fix https://github.com/zephyrproject-rtos/zephyr/issues/58325 ?

mkasenberg commented:
> @mkasenberg does this fix #58325 ?

Yes
